File: test/qiskit_basis_ordering.jl
using TEM, ITensors

let
    # Qiskit's way of writing the result.
    str_to_c(s) = complex(parse.(Float64, split(s, ","))...)
    v = open("test/qiskit_basis_ordering_final_state.dat") do f
        return str_to_c.(eachline(f))
    end

    # Now we read the OpenQASM file and build the circuit with our library.
    circuitfile = "test/qiskit_basis_ordering_circuit.qasm2"
    sites, gatelist = open(circuitfile, "r") do f
        code = read(f, String)
        gates(code, "Qubit")
    end

    psi = MPS(sites, "0")
    for g in gatelist
        psi = apply(g, psi)
    end
    cbvec = TEM.qiskitvector(psi)
    return isapprox(cbvec, v)
end

File: test/qiskit_basis_ordering.py
from qiskit import QuantumCircuit, QuantumRegister, qasm2, qasm3
import qiskit.quantum_info as qi
from math import pi

qr = QuantumRegister(3, 'q')
qc = QuantumCircuit(qr)

qc.x(qr[0])
qc.y(qr[1])
qc.h(qr[2])

print(qc.draw())

state = qi.Statevector.from_instruction(qc)

with open("qiskit_basis_ordering_final_state.dat", "w") as out:
    for x in state.data:
        print(f"{x.real},{x.imag}", file=out)

with open("qiskit_basis_ordering_circuit.qasm2", "w") as out:
    qasm2.dump(qc, out)

with open("qiskit_basis_ordering_circuit.qasm3", "w") as out:
    qasm3.dump(qc, out)
